qutebrowser/misc/sql.py

```diff
@@ -0,0 +1,244 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Provides access to an in-memory sqlite database."""
+
+import collections
+
+from PyQt5.QtCore import QObject, pyqtSignal
+from PyQt5.QtSql import QSqlDatabase, QSqlQuery
+
+from qutebrowser.utils import log
+
+
+class SqlException(Exception):
+
+    """Raised on an error interacting with the SQL database."""
+
+    pass
+
+
+def init(db_path):
+    """Initialize the SQL database connection."""
+    database = QSqlDatabase.addDatabase('QSQLITE')
+    database.setDatabaseName(db_path)
+    if not database.open():
+        raise SqlException("Failed to open sqlite database at {}: {}"
+                           .format(db_path, database.lastError().text()))
+
+
+def close():
+    """Close the SQL connection."""
+    QSqlDatabase.removeDatabase(QSqlDatabase.database().connectionName())
+
+
+def version():
+    """Return the sqlite version string."""
+    q = Query("select sqlite_version()")
+    q.run()
+    return q.value()
+
+
+class Query(QSqlQuery):
+
+    """A prepared SQL Query."""
+
+    def __init__(self, querystr, forward_only=True):
+        """Prepare a new sql query.
+
+        Args:
+            querystr: String to prepare query from.
+            forward_only: Optimization for queries that will only step forward.
+                          Must be false for completion queries.
+        """
+        super().__init__(QSqlDatabase.database())
+        log.sql.debug('Preparing SQL query: "{}"'.format(querystr))
+        if not self.prepare(querystr):
+            raise SqlException('Failed to prepare query "{}"'.format(querystr))
+        self.setForwardOnly(forward_only)
+
+    def __iter__(self):
+        if not self.isActive():
+            raise SqlException("Cannot iterate inactive query")
+        rec = self.record()
+        fields = [rec.fieldName(i) for i in range(rec.count())]
+        rowtype = collections.namedtuple('ResultRow', fields)
+
+        while self.next():
+            rec = self.record()
+            yield rowtype(*[rec.value(i) for i in range(rec.count())])
+
+    def run(self, values=None):
+        """Execute the prepared query."""
+        log.sql.debug('Running SQL query: "{}"'.format(self.lastQuery()))
+        for val in values or []:
+            self.addBindValue(val)
+        log.sql.debug('self bindings: {}'.format(self.boundValues()))
+        if not self.exec_():
+            raise SqlException('Failed to exec query "{}": "{}"'.format(
+                               self.lastQuery(), self.lastError().text()))
+        return self
+
+    def value(self):
+        """Return the result of a single-value query (e.g. an EXISTS)."""
+        if not self.next():
+            raise SqlException("No result for single-result query")
```

Comment:
Re `assert` vs `SqlException` - I see your point (programming vs. "external" errors), but I think if we already have a specialized exception type, it makes sense to use it consistently. After all, most exceptions  are programming errors - even an `SqlException` because of a failed query can be :wink: